qep_333_pyqgis_screenshots.md

```diff
@@ -0,0 +1,245 @@
+the official PyQGIS Reference Guide, specifically the screenshot component only.
+
+## Proposed Solution
+
+The old webgeodatavore site utilised manual screenshots of widgets. A manual approach
+has a number of issues:
+
+- It requires pro-active updates whenever a QGIS widget is modified and changes appearance.
+  This results in a tendency for screenshots to include outdated appearance.
+- Screenshots have a variety of styles. Some are taken on different operating systems, some
+  have different Qt themes and color schemes. This looks inconsistent and amateurish.
+- It's basically impossible for QGIS developers to add new screenshots for newly created
+  widgets with a consistent appearance as the existing images in the documentation.
+
+Accordingly, screenshots for widgets will instead be created dynamically, as part of the
+PyQGIS documentation build process. This ensures:
+
+- Widget screenshots are ALWAYS up to date with the corresponding QGIS version for the
+  documentation.
+- Widgets will ALWAYS have a consistent theme, font, sizing and general appearance.
+- It will be simple to e.g. change the Qt theme or stylesheet and have all screenshots
+  regenerated to match.
+
+To facilitate this, the [generate_docs](https://github.com/qgis/pyqgis-api-docs-builder/blob/main/scripts/make_api_rst.py#L294)
+method from [scripts/make_api_rst.py](https://github.com/qgis/pyqgis-api-docs-builder/blob/main/scripts/make_api_rst.py) will be
+modified to include an extra component in the generated class headers:
+
+``` py
+header += generate_screenshot(package, class_name, _class)
+```
+
+Where ``generate_screenshot`` is defined as follows:
+
+``` py
+def generate_screenshot(package, class_name: str, _class) -> str:
+    """
+    Generates screenshots for a class, and returns corresponding markdown
+    """
+    module_name = package.__name__.split('.')[-1]
+    script_path = Path(__file__).parent / '..' / 'screenshots' / module_name / (class_name.lower() + '.py')
+    if not script_path.exists():
+        return ""
+
+    image_path = Path(__file__).parent / '..' / 'api' / 'master' / module_name
+    spec = importlib.util.spec_from_file_location('script', script_path)
+    executed_module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(executed_module)
+    func = getattr(executed_module, 'generate_screenshot')
+    return func(image_path)
+```
+
+This function checks whether a Python file matching the current class name exists in the
+``screenshots/{module name}`` folder in the ``pyqgis-api-docs-builder`` repository. For
+example, if the current class is ``QgsMapLayerComboBox`` from the ``gui`` module, the
+function checks whether ``screenshots/gui/qgsmaplayercombobox.py`` exists. If so,
+it dynamically loads it and executes the ``generate_screenshot`` method from that file.
+
+The ``generate_screenshot`` method must have the signature:
+
+``` py
+def generate_screenshot(dest_path: Path) -> str:
```

Comment:
make this "private" using double underscores?

Reply:
Done